py/dml/c_backend.py

```diff
@@ -1093,14 +1097,145 @@ def generate_simple_events(device):
+        msg_lit = 'msg' if info.param_to_msg_comp else None
+        info.generate_callback_call(indices_lit, args_lit, msg_lit)
+        out('}\n\n', preindent = -1)
+        splitting_point()
+
+        if info.has_serialized_args:
+            start_function_definition(
+                f'attr_value_t {info.cident_args_serializer}('
+                + 'const void *_args)')
+            out('{\n', postindent = 1)
+            if info.args_type:
+                args_type_ptr = TPtr(conv_const(True, info.args_type))
+                out(f'{args_type_ptr.declaration("args")} = _args;\n')
+                args_expr = mkDereference(site, mkLit(site, 'args', args_type_ptr))
+            else:
+                args_expr = None
+            out('attr_value_t out;\n')
+            out_expr = mkLocalVariable(
+                site, LocalSymbol('out', 'out', attr_value_t, site=site))
+            info.generate_args_serializer(site, args_expr, out_expr)
+            out('return out;\n')
+            out('}\n\n', preindent = -1)
+            splitting_point()
+
+            start_function_definition(
+                f'set_error_t {info.cident_args_deserializer}('
+                + 'attr_value_t val, void *_out)')
+            out('{\n', postindent = 1)
+            out('set_error_t _success UNUSED = Sim_Set_Ok;\n')
+
+            if info.args_type:
+                out(f'{TPtr(info.args_type).declaration("out")} = _out;\n')
+                out_expr = mkDereference(site, mkLit(site, 'out',
+                                                     TPtr(info.args_type)))
+            else:
+                out_expr = None
+            def error_out(exc, msg):
+                stmts = []
+                stmts.append(mkInline(site, f'_success = {exc};'))
+                if msg is not None:
+                    stmts.append(
+                        mkInline(site, f'SIM_attribute_error("{msg}");'))
+                stmts.append(mkInline(site, 'goto _exit;'))
+                return stmts
+            val_expr = mkLit(site, 'val', attr_value_t)
+
+            info.generate_args_deserializer(site, val_expr, out_expr,
+                                            error_out)
+
+            out('_exit:\n')
+            out('return _success;\n')
+            out('}\n\n', preindent = -1)
+            splitting_point()
+
+    reset_line_directive()
+
+    if dml.globals.after_on_hook_infos:
+        items = (
+            '{%s, %s, %s, %s, %s, %d}'
+            % ((string_literal(info.string_key), info.cident_callback)
+               + ((info.cident_args_serializer, info.cident_args_deserializer)
+                  if info.has_serialized_args else ('NULL', 'NULL'))
+               + (f'sizeof({info.args_type.declaration("")})'
+                  if info.args_type else '0', info.parent.uniq))
+            for info in dml.globals.after_on_hook_infos)
+        init = '{\n%s\n}' % (',\n'.join('    %s' % (item,) for item in items),)
```

Comment:
`f'    {item}'`

Or consider merging with the previous statement.

Reply:
Agreed, either suggestion is superior to what I have currently. Not sure which one I'll pick yet though.

Reply:
Oh, no, I'll definitely pick `f'    {item}'`.

Reply:
I think you can do both! An intermediate variable with a generator is kind of odd, compare:
```
[g(item) for item in (f(x) for x in y)]
```
```
items = (f(x) for x in y)
[g(item) for item in items]
```
```
[g(item) for x in y for f(x) in x]
```
I prefer the last one: even if the nested for syntax in generator expressions is admittedly a bit weird, it's a standard python idiom and we already depend on the reader to decode them fluently in many other places.